File: server/src/Destiny/Timeline.hs
{-# LANGUAGE TemplateHaskell #-}

module Destiny.Timeline (Timeline (present), commit, modify, redo, singleton, undo, update) where

import Data.Aeson.TH (deriveJSON)
import Elm.Derive

data Timeline a = Timeline
    { past :: [a]
    , present :: a
    , future :: [a]
    , committed :: Bool
    }
deriveJSON defaultOptions ''Timeline

instance Foldable Timeline where
    foldr f z timeline = foldr f z $
        reverse (future timeline) ++ present timeline : past timeline

singleton :: a -> Timeline a
singleton x = Timeline
    { past = []
    , present = x
    , future = []
    , committed = True
    }

undo :: Timeline a -> Timeline a
undo timeline@Timeline { past = present' : past' } = timeline
    { past = past'
    , present = present'
    , future = present timeline : future timeline
    , committed = True
    }
undo timeline = timeline

redo :: Timeline a -> Timeline a
redo timeline@Timeline { future = present' : future' } = timeline
    { past = present timeline : past timeline
    , present = present'
    , future = future'
    , committed = True
    }
redo timeline = timeline

commit :: Timeline a -> Timeline a
commit timeline = timeline { committed = True }

update :: a -> Timeline a -> Timeline a
update x timeline
    | committed timeline && length timeline == maxLength =
        timeline' { past = present timeline : init (past timeline) }
    | committed timeline = timeline' { past = present timeline : past timeline }
    | otherwise = timeline'
  where
    timeline' = timeline
        { present = x
        , future = []
        , committed = False
        }
    maxLength = 50

modify :: (a -> a) -> Timeline a -> Timeline a
modify f timeline = update (f $ present timeline) timeline

File: src/Server/Timeline.fs
namespace Destiny.Server

open Destiny.Shared.Collections

type internal 'a Timeline =
    private
        { Past : 'a list
          Present : 'a
          Future : 'a list
          Committed : bool }

module internal Timeline =
    let private maxLength = 50

    let private length timeline = List.length timeline.Past + 1 + List.length timeline.Future

    let singleton value =
        { Past = []
          Present = value
          Future = []
          Committed = true }

    let present timeline = timeline.Present

    let undo = function
        | { Past = present' :: past' } as timeline ->
            { Past = past'
              Present = present'
              Future = timeline.Present :: timeline.Future
              Committed = true }
        | timeline -> timeline

    let redo = function
        | { Future = present' :: future' } as timeline ->
            { Past = timeline.Present :: timeline.Past
              Present = present'
              Future = future'
              Committed = true }
        | timeline -> timeline

    let commit timeline = { timeline with Committed = true }

    let set value timeline =
        let timeline' =
            { timeline with
                  Present = value
                  Future = []
                  Committed = false }
        if timeline.Committed && length timeline = maxLength
        then { timeline' with Past = timeline.Present :: List.initial timeline.Past }
        elif timeline.Committed
        then { timeline' with Past = timeline.Present :: timeline.Past }
        else timeline'

    let update f timeline = set (f timeline.Present) timeline
